### ISSUE TITLE
Optimize number of threads per block for each kernel

### ISSUE THREAD

**Issue by RemiLehe** [fbpic/fbpic#129]
This could be done with a standard simulation script with moving window and box full of electrons.

Even if the optimized number of threads might not generalize to any other situation, I think it will still be better than the current arbitrary choice (i.e. 256 threads per block for all kernels)

**Comment by RemiLehe** [comment]
That sounds great!! Thanks a lot for taking care of this ; I think we can really some non-negligible speed-up from this!

**Comment by soerenjalas** [comment]
I did some preliminary test on some Nvidia K40, but the results are rather inconclusive.
I used a simulation with a 2000x300 cell grid filled with particles for 1000 steps. The benchmarks scripts are attached below and can be used with the branch above.
It seems that especially with the computational heavy functions like deposition and field push there is a lot to gain from optimising the threads per block. However, it seems quite random which tpb values work well. I did the test on DESYs Maxwell cluster and it could still be that some nodes perform better than others I will test this on Jureca.

![1dtpb](https://user-images.githubusercontent.com/10922951/31225453-747be344-a9d2-11e7-8406-e16ed5f7a591.png)
![2dtpb](https://user-images.githubusercontent.com/10922951/31225455-789563b0-a9d2-11e7-9b1c-df7a779840e1.png)


[benchmark_tpb1D.zip](https://github.com/fbpic/fbpic/files/1359125/benchmark_tpb1D.zip)
[benchmark_tpb2D.zip](https://github.com/fbpic/fbpic/files/1359126/benchmark_tpb2D.zip)
